File: backend/app/routers/signatures.py
"""
Signature Routes - Signature asset management
"""
from fastapi import APIRouter, UploadFile, File, Form, HTTPException, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete
from typing import Optional
import uuid
import os
from pathlib import Path

from app.database import get_db, SignatureAsset, AuditLog

router = APIRouter()
UPLOAD_DIR = Path(__file__).parent.parent.parent / "uploads" / "signatures"
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)


@router.get("/")
async def list_signatures(status: Optional[str] = None, db: AsyncSession = Depends(get_db)):
    query = select(SignatureAsset)
    if status:
        query = query.where(SignatureAsset.status == status)
    query = query.order_by(SignatureAsset.created_at.desc())
    result = await db.execute(query)
    return result.scalars().all()


@router.get("/{signature_id}")
async def get_signature(signature_id: int, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(SignatureAsset).where(SignatureAsset.id == signature_id))
    sig = result.scalar_one_or_none()
    if not sig:
        raise HTTPException(status_code=404, detail="Signature not found")
    return sig


@router.post("/")
async def upload_signature(
    signature: UploadFile = File(...),
    uploadedBy: Optional[int] = Form(None),
    validityPeriod: str = Form("Indefinite"),
    purpose: str = Form("DL Generation"),
    adminMessage: Optional[str] = Form(None),
    db: AsyncSession = Depends(get_db)
):
    filename = f"sig_{uuid.uuid4()}{Path(signature.filename).suffix}"
    file_path = UPLOAD_DIR / filename

    content = await signature.read()
    with open(file_path, "wb") as f:
        f.write(content)

    new_sig = SignatureAsset(
        file_path=f"/uploads/signatures/{filename}",
        file_name=signature.filename,
        uploaded_by=uploadedBy,
        validity_period=validityPeriod,
        purpose=purpose,
        admin_message=adminMessage,
        status="Pending"
    )
    db.add(new_sig)
    await db.commit()
    await db.refresh(new_sig)
    return new_sig


@router.patch("/{signature_id}/approve")
async def approve_signature(signature_id: int, approvedBy: str = "Attorney", db: AsyncSession = Depends(get_db)):
    from datetime import datetime, timezone
    await db.execute(
        update(SignatureAsset)
        .where(SignatureAsset.id == signature_id)
        .values(status="Approved", approved_by=approvedBy, approved_at=datetime.now(timezone.utc))
    )
    await db.commit()
    result = await db.execute(select(SignatureAsset).where(SignatureAsset.id == signature_id))
    return result.scalar_one()


@router.patch("/{signature_id}/reject")
async def reject_signature(signature_id: int, db: AsyncSession = Depends(get_db)):
    await db.execute(update(SignatureAsset).where(SignatureAsset.id == signature_id).values(status="Rejected"))
    await db.commit()
    result = await db.execute(select(SignatureAsset).where(SignatureAsset.id == signature_id))
    return result.scalar_one()


@router.delete("/{signature_id}")
async def delete_signature(signature_id: int, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(SignatureAsset).where(SignatureAsset.id == signature_id))
    sig = result.scalar_one_or_none()
    if not sig:
        raise HTTPException(status_code=404, detail="Signature not found")

    file_path = Path(__file__).parent.parent.parent / sig.file_path.lstrip("/")
    if file_path.exists():
        os.remove(file_path)

    await db.execute(delete(SignatureAsset).where(SignatureAsset.id == signature_id))
    await db.commit()
    return {"message": "Signature deleted"}


@router.get("/status/active")
async def get_active_signature(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(SignatureAsset)
        .where(SignatureAsset.status == "Approved")
        .order_by(SignatureAsset.approved_at.desc())
        .limit(1)
    )
    return result.scalar_one_or_none()

File: server/routes/signatures.js
/**
 * Signature Routes
 * Handles signature asset management: upload, list, approve/reject, delete
 */

import express from 'express';
import multer from 'multer';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';
import { v4 as uuidv4 } from 'uuid';
import { db, logAuditAction } from '../database/db.js';
import { getManilaTimestampForSQL } from '../utils/timezone.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const router = express.Router();

// Configure multer for signature uploads
const uploadsDir = path.join(__dirname, '..', 'uploads', 'signatures');
if (!fs.existsSync(uploadsDir)) {
  fs.mkdirSync(uploadsDir, { recursive: true });
}

const storage = multer.diskStorage({
  destination: (req, file, cb) => cb(null, uploadsDir),
  filename: (req, file, cb) => {
    const uniqueName = `sig_${uuidv4()}${path.extname(file.originalname)}`;
    cb(null, uniqueName);
  }
});

const upload = multer({
  storage,
  limits: { fileSize: 5 * 1024 * 1024 }, // 5MB limit
  fileFilter: (req, file, cb) => {
    const allowedTypes = ['image/png', 'image/jpeg', 'image/webp'];
    if (allowedTypes.includes(file.mimetype)) {
      cb(null, true);
    } else {
      cb(new Error('Only PNG, JPEG, and WebP images are allowed'));
    }
  }
});

// GET /api/signatures - List all signatures
router.get('/', (req, res) => {
  try {
    const { status, userId } = req.query;
    let query = 'SELECT * FROM signature_assets WHERE 1=1';
    const params = [];

    if (status) {
      query += ' AND status = ?';
      params.push(status);
    }
    if (userId) {
      query += ' AND uploaded_by = ?';
      params.push(userId);
    }

    query += ' ORDER BY created_at DESC';
    const signatures = db.prepare(query).all(...params);
    res.json(signatures);
  } catch (error) {
    console.error('Error fetching signatures:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/signatures/:id - Get single signature
router.get('/:id', (req, res) => {
  try {
    const signature = db.prepare('SELECT * FROM signature_assets WHERE id = ?').get(req.params.id);
    if (!signature) {
      return res.status(404).json({ error: 'Signature not found' });
    }
    res.json(signature);
  } catch (error) {
    console.error('Error fetching signature:', error);
    res.status(500).json({ error: error.message });
  }
});

// POST /api/signatures - Upload new signature
router.post('/', upload.single('signature'), (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }

    const { uploadedBy, validityPeriod, purpose, adminMessage } = req.body;
    
    const stmt = db.prepare(`
      INSERT INTO signature_assets (file_path, file_name, uploaded_by, validity_period, purpose, admin_message, status)
      VALUES (?, ?, ?, ?, ?, ?, 'Pending')
    `);
    
    const result = stmt.run(
      `/uploads/signatures/${req.file.filename}`,
      req.file.originalname,
      uploadedBy || null,
      validityPeriod || 'Indefinite',
      purpose || 'DL Generation',
      adminMessage || null
    );

    // Log audit action
    logAuditAction(
      uploadedBy || null,
      'Admin',
      'SIGNATURE_UPLOADED',
      'signature_assets',
      result.lastInsertRowid,
      { fileName: req.file.originalname }
    );

    const newSignature = db.prepare('SELECT * FROM signature_assets WHERE id = ?').get(result.lastInsertRowid);
    res.status(201).json(newSignature);
  } catch (error) {
    console.error('Error uploading signature:', error);
    res.status(500).json({ error: error.message });
  }
});

// PATCH /api/signatures/:id/approve - Approve signature
router.patch('/:id/approve', (req, res) => {
  try {
    const { approvedBy, reason } = req.body;
    
    // FIXED: Use Manila timezone timestamp instead of CURRENT_TIMESTAMP (which is UTC)
    const manilaTimestamp = getManilaTimestampForSQL();
    
    db.prepare(`
      UPDATE signature_assets 
      SET status = 'Approved', approved_by = ?, approved_at = ?, updated_at = ?
      WHERE id = ?
    `).run(approvedBy || 'Attorney', manilaTimestamp, manilaTimestamp, req.params.id);

    logAuditAction(null, approvedBy, 'SIGNATURE_APPROVED', 'signature_assets', req.params.id, { reason });

    const signature = db.prepare('SELECT * FROM signature_assets WHERE id = ?').get(req.params.id);
    res.json(signature);
  } catch (error) {
    console.error('Error approving signature:', error);
    res.status(500).json({ error: error.message });
  }
});

// PATCH /api/signatures/:id/reject - Reject signature
router.patch('/:id/reject', (req, res) => {
  try {
    const { rejectedBy, reason } = req.body;
    
    // FIXED: Use Manila timezone timestamp instead of CURRENT_TIMESTAMP (which is UTC)
    const manilaTimestamp = getManilaTimestampForSQL();
    
    db.prepare(`
      UPDATE signature_assets 
      SET status = 'Rejected', updated_at = ?
      WHERE id = ?
    `).run(manilaTimestamp, req.params.id);

    logAuditAction(null, rejectedBy, 'SIGNATURE_REJECTED', 'signature_assets', req.params.id, { reason });

    const signature = db.prepare('SELECT * FROM signature_assets WHERE id = ?').get(req.params.id);
    res.json(signature);
  } catch (error) {
    console.error('Error rejecting signature:', error);
    res.status(500).json({ error: error.message });
  }
});

// DELETE /api/signatures/:id - Delete signature
router.delete('/:id', (req, res) => {
  try {
    const signature = db.prepare('SELECT * FROM signature_assets WHERE id = ?').get(req.params.id);
    if (!signature) {
      return res.status(404).json({ error: 'Signature not found' });
    }

    // Delete file
    const filePath = path.join(__dirname, '..', signature.file_path);
    if (fs.existsSync(filePath)) {
      fs.unlinkSync(filePath);
    }

    db.prepare('DELETE FROM signature_assets WHERE id = ?').run(req.params.id);
    
    logAuditAction(null, 'Admin', 'SIGNATURE_DELETED', 'signature_assets', req.params.id, { fileName: signature.file_name });

    res.json({ message: 'Signature deleted successfully' });
  } catch (error) {
    console.error('Error deleting signature:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/signatures/active - Get active approved signature
router.get('/status/active', (req, res) => {
  try {
    const signature = db.prepare(`
      SELECT * FROM signature_assets 
      WHERE status = 'Approved' 
      ORDER BY approved_at DESC 
      LIMIT 1
    `).get();
    
    res.json(signature || null);
  } catch (error) {
    console.error('Error fetching active signature:', error);
    res.status(500).json({ error: error.message });
  }
});

export { router as signatureRoutes };
